lib/pf/provisioner/google_workspace_chromebook.pm

```diff
@@ -0,0 +1,465 @@
+    for my $device (@$devices) {
+        $self->process_device($device);
+    }
+}
+
+sub cache {
+    return pf::CHI->new(namespace => 'provisioning_distributed');
+}
+
+sub syncQueryKey {
+    my ($self) = @_;
+    return $self->id . ":syncQuery";
+}
+
+sub syncQuery {
+    my ($self) = @_;
+    my $syncQuery = $self->cache->get($self->syncQueryKey ) // "sync:0000-01-02..";
+    return $syncQuery;
+}
+
+sub authorize {
+    my ($self, $mac) = @_;
+    my ($err, $device) = $self->getDevice($mac);
+    if (defined $err) {
+        return $err;
+    }
+
+    if ($device->{status} eq 'ACTIVE') {
+        return $TRUE;
+    }
+
+    return $FALSE;
+}
+
+sub pollAndEnforce {
+    my ($self, $timeframe) = @_;
+    my $nextToken;
+    my $query = $self->syncQuery;
+    my $chromeosdevices;
+    while (1) {
+        my ( $err, $list ) = $self->getList(
+            {
+                pageToken => $nextToken,
+                orderBy   => 'LAST_SYNC',
+                query     => $query,
+            }
+        );
+        if (defined $err) {
+            $logger->error("Unable to contact the Google API to poll the changed devices.");
+            return $err;
+        }
+
+        $chromeosdevices = $list->{chromeosdevices} // [];
+        $self->process_devices($chromeosdevices);
+        $nextToken = $list->{nextPageToken};
+        last if !defined $nextToken;
+    }
+
+    if (defined $chromeosdevices && @$chromeosdevices) {
+        $self->cache->set($self->syncQueryKey, syncDateToQuery($chromeosdevices->[-1]->{lastSync}));
+    }
+}
+
+sub getNextImportDevices {
+    my ($self, $nextToken) = @_;
+    return $self->getList(
+        {
+            pageToken => $nextToken,
+            orderBy   => 'LAST_SYNC',
+            query     => $self->importLastSyncQuery()
+        }
+    );
+}
+
+sub importDevices {
+    my ($self) = @_;
+    my $nextToken;
+    my $query = $self->importLastSyncQuery();
+    while (1) {
+        my ($err, $list) = $self->getNextImportDevices($nextToken);
+        if (defined $err) {
+            $logger->error("Unable to contact the Google API to import new devices.");
+            return $err;
+        }
+
+        my $chromeosdevices = $list->{chromeosdevices} // [];
+        $self->_import_devices($chromeosdevices);
+        $nextToken = $list->{nextPageToken};
+        last if !defined $nextToken;
+    }
+}
+
+sub importLastSyncQuery {
+    my ($self) = @_;
+    return "sync:0000-01-01..";
+}
+
+sub _import_devices {
+    my ($self, $chromeosdevices) = @_;
+    for my $device (@$chromeosdevices) {
+        $self->_import_device($device);
+    }
+}
+
+sub _import_device {
+    my ($self, $device) = @_;
+    return if $device->{status} ne 'ACTIVE';
+    my $role = $self->role_to_apply;
+    for my $f (qw(macAddress ethernetMacAddress ethernetMacAddress0)) {
+        next if !exists $device->{$f};
+        my $mac = $device->{$f};
+        next if !defined $mac;
+        node_add($mac, category => $role );
```

Comment:
Shouldn't it be node_register ? Otherwise it imports it in PF but it won't be registered AFAIK